File: uni-idea-plugin/src/main/java/com/intellij/ide/util/treeView/AbstractTreeBuilder2.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package com.intellij.ide.util.treeView;

import com.intellij.my.file.AbstractProjectTreeStructure2;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.Application;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.Progressive;
import com.intellij.openapi.util.ActionCallback;
import com.intellij.psi.impl.smartPointers.NodeDescriptor2;
import com.intellij.reference.SoftReference;
import com.intellij.util.Consumer;
import com.intellij.util.concurrency.EdtExecutorService;
import com.intellij.util.ui.UIUtil;
import com.intellij.util.ui.update.MergingUpdateQueue;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.concurrency.AsyncPromise;
import org.jetbrains.concurrency.Promise;
import org.jetbrains.concurrency.Promises;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import java.lang.ref.Reference;
import java.lang.ref.WeakReference;
import java.util.*;

/**
 * @deprecated use {@link com.intellij.ui.tree.AsyncTreeModel} and {@link com.intellij.ui.tree.StructureTreeModel} instead.
 */
@SuppressWarnings("UnstableApiUsage")
@ApiStatus.ScheduledForRemoval(inVersion = "2020.3")
@Deprecated
public class AbstractTreeBuilder2 implements Disposable {
  private AbstractTreeUi2 myUi;
  @NonNls private static final String TREE_BUILDER = "TreeBuilder";
  protected static final boolean DEFAULT_UPDATE_INACTIVE = true;

  protected AbstractTreeBuilder2() {

  }

  protected void init(@NotNull JTree tree,
                      @NotNull DefaultTreeModel treeModel,
                      AbstractProjectTreeStructure2 treeStructure) {

    tree.putClientProperty(TREE_BUILDER, new WeakReference<>(this));

    myUi = createUi();
    myUi.init(this, tree, treeModel, treeStructure);

    setPassthroughMode(isUnitTestingMode());
  }

  @NotNull
  protected AbstractTreeUi2 createUi() {
    return new AbstractTreeUi2();
  }

  public final void scrollTo(Object element, Runnable onDone) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.userScrollTo(element, onDone == null ? null : new UserRunnable(onDone));
  }

  public final void select(final Object element, @Nullable final Runnable onDone) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.userSelect(new Object[]{element}, new UserRunnable(onDone), false, true);
  }

  public final void expand(Object element, @Nullable Runnable onDone) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.expand(element, new UserRunnable(onDone));
  }

  public final void collapseChildren(Object element, @Nullable Runnable onDone) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.collapseChildren(element, new UserRunnable(onDone));
  }


  @Nullable
  protected AbstractTreeUpdater2 createUpdater() {
    if (isDisposed()) return null;

    AbstractTreeUpdater2 updater = new AbstractTreeUpdater2(this);
    updater.setModalityStateComponent(MergingUpdateQueue.ANY_COMPONENT);
    return updater;
  }

  @Nullable
  protected final AbstractTreeUpdater2 getUpdater() {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.getUpdater();
  }

  public final boolean addSubtreeToUpdateByElement(@NotNull Object element) {
    AbstractTreeUpdater2 updater = getUpdater();
    return updater != null && updater.addSubtreeToUpdateByElement(element);
  }

  @Nullable
  public final DefaultMutableTreeNode getRootNode() {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.getRootNode();
  }

  public final void setNodeDescriptorComparator(Comparator<? super NodeDescriptor2<?>> nodeDescriptorComparator) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) {
      ui.setNodeDescriptorComparator(nodeDescriptorComparator);
    }
  }

  /**
   * node descriptor getElement contract is as follows:
   * 1.TreeStructure always returns & receives "treeStructure" element returned by getTreeStructureElement
   * 2.Paths contain "model" element returned by getElement
   */
  protected Object getTreeStructureElement(NodeDescriptor2 nodeDescriptor) {
    return nodeDescriptor == null ? null : nodeDescriptor.getElement();
  }


  protected boolean validateNode(@NotNull Object child) {
    AbstractTreeStructure structure = getTreeStructure();
    return structure != null && structure.isValid(child);
  }

  public final JTree getTree() {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.getTree();
  }

  public final AbstractTreeStructure getTreeStructure() {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.getTreeStructure();
  }

  @Nullable
  public Object getRootElement() {
    AbstractTreeStructure structure = getTreeStructure();
    return structure == null ? null : structure.getRootElement();
  }

  /**
   * @deprecated use {@link #queueUpdate()}
   */
  @Deprecated
  public void updateFromRoot() {
    queueUpdate();
  }

  public void initRootNode() {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.initRootNode();
  }

  @NotNull
  public final ActionCallback queueUpdate() {
    return queueUpdate(true);
  }

  @NotNull
  public final ActionCallback queueUpdate(boolean withStructure) {
    return queueUpdateFrom(getRootElement(), true, withStructure);
  }

  public final void queueUpdateFrom(final Object element, final boolean forceResort) {
    queueUpdateFrom(element, forceResort, true);
  }

  @NotNull
  public ActionCallback queueUpdateFrom(final Object element, final boolean forceResort, final boolean updateStructure) {
    AbstractTreeUi2 ui = getUi();
    if (ui == null) {
      return ActionCallback.REJECTED;
    }

    final ActionCallback result = new ActionCallback();
    ui.invokeLaterIfNeeded(false, new TreeRunnable2("AbstractTreeBuilder.queueUpdateFrom") {
      @Override
      public void perform() {
        AbstractTreeUi2 ui = getUi();
        if (ui == null) {
          result.reject("ui is null");
          return;
        }

        if (updateStructure && forceResort) {
          ui.incComparatorStamp();
        }
        ui.queueUpdate(element, updateStructure).notify(result);
      }
    });
    return result;
  }

  /**
   * @deprecated use {@link AbstractTreeUi#getNodeForElement(Object, boolean)}
   */
  @Deprecated
  @Nullable
  public DefaultMutableTreeNode getNodeForElement(@NotNull Object element) {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.getNodeForElement(element, false);
  }

  public void cleanUp() {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.doCleanUp();
  }

  @Nullable
  protected ProgressIndicator createProgressIndicator() {
    return null;
  }

  protected void expandNodeChildren(@NotNull DefaultMutableTreeNode node) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.doExpandNodeChildren(node);
  }

  protected boolean isAutoExpandNode(final NodeDescriptor2 nodeDescriptor) {
    return !isDisposed() && getRootElement() == getTreeStructureElement(nodeDescriptor);
  }

  protected boolean isAlwaysShowPlus(final NodeDescriptor2 descriptor) {
    return false;
  }


  protected boolean isSmartExpand() {
    return true;
  }

  public final boolean isDisposed() {
    return getUi() == null;
  }

  @Nullable
  protected Object findNodeByElement(@NotNull Object element) {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? null : ui.findNodeByElement(element);
  }

  void runOnYieldingDone(@NotNull Runnable onDone) {
    AbstractTreeUi2 ui = getUi();
    if (ui == null) return;

    if (ui.isPassthroughMode() || SwingUtilities.isEventDispatchThread()) {
      onDone.run();
    }
    else {
      EdtExecutorService.getInstance().execute(()->{
        if (!isDisposed()) onDone.run();
      });
    }
  }

  protected void yieldToEDT(@NotNull Runnable runnable) {
    AbstractTreeUi2 ui = getUi();
    if (ui == null) return;

    if (ui.isPassthroughMode()) {
      runnable.run();
    }
    else {
      EdtExecutorService.getInstance().execute(()->{
        if (!isDisposed()) runnable.run();
      });
    }
  }

  public boolean isToEnsureSelectionOnFocusGained() {
    return true;
  }

  protected void runBackgroundLoading(@NotNull final Runnable runnable) {
    if (isDisposed()) return;

    final Application app = ApplicationManager.getApplication();
    if (app != null) {
      app.runReadAction(new TreeRunnable2("AbstractTreeBuilder.runBackgroundLoading") {
        @Override
        public void perform() {
          runnable.run();
        }
      });
    }
    else {
      runnable.run();
    }
  }

  protected void updateAfterLoadedInBackground(@NotNull Runnable runnable) {
    AbstractTreeUi2 ui = getUi();
    if (ui == null) return;

    if (ui.isPassthroughMode()) {
      runnable.run();
    }
    else {
      UIUtil.invokeLaterIfNeeded(runnable);
    }
  }

  @NotNull
  public final ActionCallback getReady(Object requestor) {
    AbstractTreeUi2 ui = getUi();
    return ui == null ? ActionCallback.REJECTED : ui.getReady(requestor);
  }

  public void setPassthroughMode(boolean passthrough) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.setPassthroughMode(passthrough);
  }

  public void batch(@NotNull Progressive progressive) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) {
      ui.batch(progressive);
    }
  }

  @NotNull
  public Promise<Object> revalidateElement(@NotNull Object element) {
    AbstractTreeStructure structure = getTreeStructure();
    if (structure == null) {
      return Promises.rejectedPromise();
    }

    AsyncPromise<Object> promise = new AsyncPromise<>();
    structure
      .revalidateElement(element)
      .doWhenDone((Consumer<Object>)o -> promise.setResult(o))
      .doWhenRejected(s -> promise.setError(s));
    return promise;
  }

  public final AbstractTreeUi2 getUi() {
    AbstractTreeUi2 ui = myUi;
    return ui == null || ui.isReleaseRequested() ? null : ui;
  }

  @Override
  public void dispose() {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.requestRelease();
  }

  void releaseUi() {
    myUi = null;
  }

  protected boolean updateNodeDescriptor(@NotNull NodeDescriptor2 descriptor) {
    AbstractTreeUi2 ui = getUi();
    return ui != null && descriptor.update();
  }

  public final void setCanYieldUpdate(boolean yield) {
    AbstractTreeUi2 ui = getUi();
    if (ui != null) ui.setCanYield(yield);
  }

  @Nullable
  public static AbstractTreeBuilder2 getBuilderFor(@NotNull JTree tree) {
    Reference<AbstractTreeBuilder2> ref = (Reference)tree.getClientProperty(TREE_BUILDER);
    return SoftReference.dereference(ref);
  }

  private static boolean isUnitTestingMode() {
    Application app = ApplicationManager.getApplication();
    return app != null && app.isUnitTestMode();
  }

  public static boolean isToPaintSelection(@NotNull JTree tree) {
    AbstractTreeBuilder2 builder = getBuilderFor(tree);
    return builder == null || builder.getUi() == null || builder.getUi().isToPaintSelection();
  }

  class UserRunnable implements Runnable {
    private final Runnable myRunnable;

    UserRunnable(Runnable runnable) {
      myRunnable = runnable;
    }

    @Override
    public void run() {
      if (myRunnable != null) {
        AbstractTreeUi2 ui = getUi();
        if (ui != null) {
          ui.executeUserRunnable(myRunnable);
        }
        else {
          myRunnable.run();
        }
      }
    }
  }

}

File: uni-idea-plugin/src/main/java/com/intellij/ide/util/treeView/AbstractTreeBuilder2.kt
package com.intellij.ide.util.treeView

import com.intellij.ide.projectView.impl.nodes.AbstractTreeNod2
import com.intellij.my.file.AbstractProjectTreeStructure2
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.Progressive
import com.intellij.openapi.util.ActionCallback
import com.intellij.reference.SoftReference
import com.intellij.util.Consumer
import com.intellij.util.concurrency.EdtExecutorService
import com.intellij.util.ui.UIUtil
import com.intellij.util.ui.update.MergingUpdateQueue
import org.jetbrains.annotations.NonNls
import org.jetbrains.concurrency.AsyncPromise
import org.jetbrains.concurrency.Promise
import org.jetbrains.concurrency.rejectedPromise
import java.lang.ref.Reference
import java.lang.ref.WeakReference
import javax.swing.JTree
import javax.swing.SwingUtilities
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel

@Deprecated("use {@link com.intellij.ui.tree.AsyncTreeModel} and {@link com.intellij.ui.tree.StructureTreeModel} instead.")
abstract class AbstractTreeBuilder2 protected constructor() : Disposable {
    private var myUi: AbstractTreeUi2? = null
    protected fun init(
        tree: JTree,
        treeModel: DefaultTreeModel,
        treeStructure: AbstractProjectTreeStructure2?
    ) {
        tree.putClientProperty(TREE_BUILDER, WeakReference(this))
        myUi = createUi()
        myUi!!.init(this, tree, treeModel, treeStructure)
        setPassthroughMode(isUnitTestingMode)
    }

    protected fun createUi(): AbstractTreeUi2 {
        return AbstractTreeUi2()
    }

    fun scrollTo(element: Any?, onDone: Runnable?) {
        val ui = ui
        ui?.userScrollTo(
            element,
            if (onDone == null) null else UserRunnable(onDone)
        )
    }

    fun expand(element: Any?, onDone: Runnable?) {
        val ui = ui
        ui?.expand(element, UserRunnable(onDone))
    }

    fun collapseChildren(element: Any?, onDone: Runnable?) {
        val ui = ui
        ui?.collapseChildren(element!!, UserRunnable(onDone))
    }

    fun createUpdater(): AbstractTreeUpdater2? {
        if (isDisposed) return null
        val updater = AbstractTreeUpdater2(this)
        updater.setModalityStateComponent(MergingUpdateQueue.ANY_COMPONENT)
        return updater
    }

    protected val updater: AbstractTreeUpdater2?
        protected get() {
            val ui = ui
            return ui?.updater
        }

    fun addSubtreeToUpdateByElement(element: Any): Boolean {
        val updater = updater
        return updater != null && updater.addSubtreeToUpdateByElement(element)
    }

    val rootNode: DefaultMutableTreeNode?
        get() {
            val ui = ui
            return ui?.rootNode
        }

    fun setNodeDescriptorComparator(nodeDescriptorComparator: Comparator<in AbstractTreeNod2<*>?>?) {
        val ui = ui
        ui?.setNodeDescriptorComparator(nodeDescriptorComparator)
    }

    /**
     * node descriptor getElement contract is as follows:
     * 1.TreeStructure always returns & receives "treeStructure" element returned by getTreeStructureElement
     * 2.Paths contain "model" element returned by getElement
     */
    fun getTreeStructureElement(nodeDescriptor: AbstractTreeNod2<*>?): AbstractTreeNod2<*>? {
        return nodeDescriptor?.element
    }

    open fun validateNode(child: Any): Boolean {
        val structure = treeStructure
        return structure != null && structure.isValid(child)
    }

    val tree: JTree?
        get() {
            val ui = ui
            return ui?.tree
        }
    val treeStructure: AbstractTreeStructure?
        get() {
            val ui = ui
            return ui?.treeStructure
        }
    val rootElement: Any?
        get() {
            val structure = treeStructure
            return structure?.rootElement
        }

    @Deprecated("use {@link #queueUpdate()}")
    fun updateFromRoot() {
        queueUpdate()
    }

    fun initRootNode() {
        val ui = ui
        ui?.initRootNode()
    }

    @JvmOverloads
    fun queueUpdate(withStructure: Boolean = true): ActionCallback {
        return queueUpdateFrom(rootElement, true, withStructure)
    }

    fun queueUpdateFrom(element: Any?, forceResort: Boolean) {
        queueUpdateFrom(element, forceResort, true)
    }

    fun queueUpdateFrom(element: Any?, forceResort: Boolean, updateStructure: Boolean): ActionCallback {
        val ui = ui ?: return ActionCallback.REJECTED
        val result = ActionCallback()
        ui.invokeLaterIfNeeded(false, object : TreeRunnable2("AbstractTreeBuilder.queueUpdateFrom") {
            public override fun perform() {
              if (updateStructure && forceResort) {
                    ui.incComparatorStamp()
                }
                ui.queueUpdate(element, updateStructure).notify(result)
            }
        })
        return result
    }

    @Deprecated("")
    fun getNodeForElement(element: Any): DefaultMutableTreeNode? {
        val ui = ui
        return ui?.getNodeForElement(element, false)
    }

    fun cleanUp() {
        val ui = ui
        ui?.doCleanUp()
    }

    open fun createProgressIndicator(): ProgressIndicator? {
        return null
    }

    open fun expandNodeChildren(node: DefaultMutableTreeNode) {
        val ui = ui
        ui?.doExpandNodeChildren(node)
    }

    abstract fun isAlwaysShowPlus(descriptor: AbstractTreeNod2<*>?): Boolean
    val isDisposed: Boolean
        get() = ui == null

    fun findNodeByElement(element: Any): Any? {
        val ui = ui
        return ui?.findNodeByElement(element)
    }

    fun runOnYieldingDone(onDone: Runnable) {
        val ui = ui ?: return
        if (ui.isPassthroughMode || SwingUtilities.isEventDispatchThread()) {
            onDone.run()
        } else {
            EdtExecutorService.getInstance().execute { if (!isDisposed) onDone.run() }
        }
    }

    fun yieldToEDT(runnable: Runnable) {
        val ui = ui ?: return
        if (ui.isPassthroughMode) {
            runnable.run()
        } else {
            EdtExecutorService.getInstance().execute { if (!isDisposed) runnable.run() }
        }
    }

    fun runBackgroundLoading(runnable: Runnable) {
        if (isDisposed) return
        val app = ApplicationManager.getApplication()
        if (app != null) {
            app.runReadAction(object : TreeRunnable2("AbstractTreeBuilder.runBackgroundLoading") {
                public override fun perform() {
                    runnable.run()
                }
            })
        } else {
            runnable.run()
        }
    }

    fun updateAfterLoadedInBackground(runnable: Runnable) {
        val ui = ui ?: return
        if (ui.isPassthroughMode) {
            runnable.run()
        } else {
            UIUtil.invokeLaterIfNeeded(runnable)
        }
    }

    fun getReady(requestor: Any?): ActionCallback {
        val ui = ui
        return if (ui == null) ActionCallback.REJECTED else ui.getReady(requestor!!)
    }

    fun setPassthroughMode(passthrough: Boolean) {
        val ui = ui
        if (ui != null) ui.isPassthroughMode = passthrough
    }

    fun batch(progressive: Progressive) {
        val ui = ui
        ui?.batch(progressive)
    }

    open fun revalidateElement(element: Any): Promise<Any?> {
        val structure = treeStructure ?: return rejectedPromise()
        val promise = AsyncPromise<Any?>()
        structure
            .revalidateElement(element)
            .doWhenDone(Consumer { o: Any? -> promise.setResult(o) })
            .doWhenRejected { s: String? -> promise.setError(s!!) }
        return promise
    }

    val ui: AbstractTreeUi2?
        get() {
            val ui = myUi
            return if (ui == null || ui.isReleaseRequested) null else ui
        }

    override fun dispose() {
        val ui = ui
        ui?.requestRelease()
    }

    fun releaseUi() {
        myUi = null
    }

    fun updateNodeDescriptor(descriptor: AbstractTreeNod2<*>): Boolean {
        val ui = ui
        return ui != null && descriptor.update()
    }

    fun setCanYieldUpdate(yield2: Boolean) {
        val ui = ui
        ui?.setCanYield(yield2)
    }

    internal inner class UserRunnable(private val myRunnable: Runnable?) : Runnable {
        override fun run() {
            if (myRunnable != null) {
                val ui = ui
                if (ui != null) {
                    ui.executeUserRunnable(myRunnable)
                } else {
                    myRunnable.run()
                }
            }
        }
    }

    companion object {
        @NonNls
        private val TREE_BUILDER = "TreeBuilder"
        const val DEFAULT_UPDATE_INACTIVE = true

      private val isUnitTestingMode: Boolean
            private get() {
                val app = ApplicationManager.getApplication()
                return app != null && app.isUnitTestMode
            }
    }
}